### PR TITLE
Firefly-995: support defining then passing firefly options from server to client

### PR DESCRIPTION
**Pull request by robyww**
#### [Firefly-995](https://jira.ipac.caltech.edu/browse/FIREFLY-995)

  - The property java `FIREFLY_OPTIONS` can contain a json string equivalent  `firefly.options` on the client. 
  - This json object will be merge with what is running on the client at startup time.
  - Also fixes a bug I introduced in my last PR that breaks the advanced TAP panel. Related to converting `InputAreaFieldView` to a functional component

#### Testing
  - use docker to test
  - define an environment variable from bash
     - `export PROPS_FIREFLY_OPTIONS=$'\'{ "tap" : { "additional":{ "services": [ { "label": "I-internal", "value": "https://irsasearchops.ipac.caltech.edu/TAP" } ] } } }\''`
  - `docker compose up firefly`
  - The tap option should show a new tap service labeled 'I-internal'
  - To override the IRSA server try: 
    - `export PROPS_FIREFLY_OPTIONS=$'\'{ "tap" : { "additional":{ "services": [ { "label": "IRSA", "value": "https://irsasearchops.ipac.caltech.edu/TAP" } ] } } }\''`

### CODE REVIEW

File: docker-compose.yml [hunk 1]
```diff
@@ -9,6 +9,7 @@ services:
       - "8080:8080"
     environment: # any variable listed in catalina.sh and usually set in tomcat's setenv.sh, could be set here
       - ADMIN_PASSWORD      # value taken from env
+      - PROPS_FIREFLY_OPTIONS
```
Comment (loitly) [inline docker-compose.yml:12]: Just wanted to point out...
Any firefly props can be sent in via this method.  The ones declared here are either required or highly recommended when running Firefly.

File: src/firefly/js/Firefly.js [hunk 2]
```diff
@@ -323,17 +323,29 @@ function bootstrap(props, options, webApiCommands) {
...
+        catch (err) {
+            logger.error('could not retrieve valid server options');
+        }
+
+        const finalOptions = mergeObjectOnly(options, srvOptions);
+        const client= await getOrCreateWsConn();
+
+        fireflyInit(props, finalOptions, webApiCommands);
+
+        client.addListener(ActionEventHandler);
```
Comment (loitly) [inline src/firefly/js/Firefly.js:340]: Nice.  Much cleaner.

File: src/firefly/js/ui/tap/TapUtil.js [hunk 3]
```diff
@@ -218,13 +218,24 @@ export function getColumnAttribute(columnsModel, colName, attrName) {
...
 
+function mergeAdditionalServices(tapServices, additional) {
+    if (!hasElements(additional)) return tapServices;
+
+    const modifiedOriginal= tapServices.map( (t) => {
+        const match= additional.find( (a) => a.label===t.label);
+        return match ? {...t,...match} : t;
+    });
+    const trulyAdditional= additional.filter( (a) => tapServices.filter( (t) => t.label===a.label).length===0 );
+
```
Comment (loitly) [inline src/firefly/js/ui/tap/TapUtil.js:229]: Would a `tapServices.find()` be more appropriate here?

File: src/firefly/js/ui/tap/TapUtil.js [hunk 4]
```diff
@@ -218,13 +218,24 @@ export function getColumnAttribute(columnsModel, colName, attrName) {
...
 export function getTapServices(webApiUserAddedService) {
-    const tapServices = getAppOptions()?.tap?.services;
-    const additionalServices = getAppOptions()?.tap?.additional?.services;
-    const retVal= hasElements(tapServices) ? [...tapServices] : [...TAP_SERVICES_FALLBACK];
-    hasElements(additionalServices) && retVal.unshift(...additionalServices);
-    webApiUserAddedService && retVal.push(webApiUserAddedService);
-    return retVal;
+    const {tap} = getAppOptions();
+    const startingTapServices= hasElements(tap?.services) ? [...tap.services] : [...TAP_SERVICES_FALLBACK];
+    const mergedServices= mergeAdditionalServices(startingTapServices,tap?.additional?.services);
```
Comment (loitly) [inline src/firefly/js/ui/tap/TapUtil.js:236]: For the purpose of adding additional services to the top of the list, this works fine.  However, a more elaborate scheme is needed to truly manage the list, like insert and remove.

**Review by loitly — APPROVED** [review]
Tested as instructed and it worked fine.  Good to go.

File: docker-compose.yml (hunk 1 above)
Comment (robyww) [reply docker-compose.yml:12]: I will remove  `PROPS_FIREFLY_OPTIONS` here

File: src/firefly/js/ui/tap/TapUtil.js (hunk 3 above)
Comment (robyww) [reply src/firefly/js/ui/tap/TapUtil.js:229]: yea, I should use `find`

**Pull request merged**